### PR TITLE
Decentralize parachain relayer with linear timeout

### PR DESCRIPTION
**Pull request by yrong**
### Context

This is the implementation [idea](https://github.com/Snowfork/snowbridge/pull/1265#discussion_r1715433869) from Alistair, essentially linear waiting time for each relayer to avoid race condition, it's more deterministic/predictable.

So basically we've 2 options

- random polling + fixed waiting time in https://github.com/Snowfork/snowbridge/pull/1265
- subscribe + linear increased waiting time

The PR here for option 2.

### CODE REVIEW

File: relayer/relays/parachain/config.go [hunk 1]
```diff
@@ -32,6 +35,23 @@ type SinkContractsConfig struct {
 	Gateway string `mapstructure:"Gateway"`
 }
 
+type ScheduleConfig struct {
+	// ID of current relayer, starting from 0
+	ID uint64 `mapstructure:"id"`
+	// Number of total count of all relayers
+	Num uint64 `mapstructure:"num"`
```
Comment (vgeddes) [inline relayer/relays/parachain/config.go:42]:
```suggestion
	TotalRelayerCount uint64 `mapstructure:"totalRelayerCount"`
```

File: relayer/relays/parachain/beefy-listener.go [hunk 2]
```diff
@@ -154,18 +158,18 @@ func (li *BeefyListener) doScan(ctx context.Context, beefyBlockNumber uint64) er
...
-		if err != nil {
-			return err
+		paraNonce := (*task.MessageProofs)[0].Message.Nonce
+		modNonce := paraNonce % li.scheduleConfig.Num
+		var waitingPeriod uint64
+		if modNonce > li.scheduleConfig.ID {
+			waitingPeriod = modNonce - li.scheduleConfig.ID
+		} else {
+			waitingPeriod = li.scheduleConfig.ID - modNonce
 		}
```
Comment (alistair-singh) [inline relayer/relays/parachain/beefy-listener.go:169]: Please double check but the math below will produce the waiting period correctly for all input.
```suggestion
		paraNonce := (*task.MessageProofs)[0].Message.Nonce
	        waitingPeriod := (paraNonce + li.scheduleConfig.Num - li.scheduleConfig.ID) % li.scheduleConfig.Num
```

Comment (yrong) [reply relayer/relays/parachain/beefy-listener.go]: Nice catch! Fixed in https://github.com/Snowfork/snowbridge/pull/1266/commits/e787143109cf5d4b71b9afb84c1acb20892b6a35

File: relayer/relays/parachain/config.go (hunk 1 above)
Comment (yrong) [reply relayer/relays/parachain/config.go]: https://github.com/Snowfork/snowbridge/pull/1266/commits/0d648c91a95a788296b13ac14d5e3dacf7945f59

File: relayer/relays/parachain/beefy-listener.go [hunk 3]
```diff
@@ -285,3 +289,37 @@ func (li *BeefyListener) generateProof(ctx context.Context, input *ProofInput, h
...
+			return err
+		}
+		if ethInboundNonce >= paraNonce {
+			log.Info(fmt.Sprintf("nonce %d picked up by another relayer, just skip", paraNonce))
+			return nil
+		}
+		if cnt == waitingPeriod {
+			break
+		}
+		time.Sleep(45 * time.Second)
```
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/parachain/beefy-listener.go]: https://github.com/Snowfork/snowbridge/pull/1266/commits/1ced891a5fee9ea849f3a4fdc6327b17cea86814

**Review by alistair-singh — APPROVED** [review]

**Comment by acatangiu** [comment]
Just to make sure I'm on the same page:

These changes are a "best effort" decentralization that assumes all relayers willingly follow this protocol and wait for their turn, right?

IIUC this will help with deploying multiple instances of these "nice/behaving" relayers, but there is no mechanism that prevents a relayer looking to maximize gains to just step on the other's slots.

**Pull request merged**